### ISSUE TITLE
PHP 8.1: strlen(): Passing null to parameter #1 ($string) of type string is deprecated in backdrop_random_bytes()

### ISSUE THREAD

**Issue by kiamlaluno** [backdrop/backdrop-issues#5727]
## Description of the bug

[`backdrop_random_bytes()`](https://docs.backdropcms.org/api/backdrop/core!includes!bootstrap.inc/function/backdrop_random_bytes/1) uses the following code.

```php
    if (strlen($bytes) < $count) {
      // Initialize on the first call. The contents of $_SERVER includes a mix of
      // user-specific and system information that varies a little with each page.
      if (!isset($random_state)) {
        $random_state = print_r($_SERVER, TRUE);
        if (function_exists('getmypid')) {
          // Further initialize with the somewhat random PHP process ID.
          $random_state .= getmypid();
        }
        $bytes = '';
      }

      do {
        $random_state = hash('sha256', microtime() . mt_rand() . $random_state);
        $bytes .= hash('sha256', mt_rand() . $random_state, TRUE);
      } 
       while (strlen($bytes) < $count);
    }
```

It passes the static `$bytes` variable, which could still be uninitialized, to `strlen()`.

In PHP 8.1, passing `NULL` to `strlen()` throws an exception.

## Expected behavior
As done in one of the first function lines, the code should cast `$bytes` to `string` in the `strlen()` calls.

**Comment by indigoxela** [comment]
> In PHP 8.1, passing NULL to strlen() throws an exception

That's right, but do you have any real example to end up with the exception? Steps to reproduce, ideally.

**Comment by kiamlaluno** [comment]
The code before the code I quoted is the following.

```php
    // openssl_random_pseudo_bytes() will find entropy in a system-dependent
    // way.
    if (function_exists('openssl_random_pseudo_bytes')) {
      $bytes .= openssl_random_pseudo_bytes($missing_bytes);
    }

    // Else, read directly from /dev/urandom, which is available on many *nix
    // systems and is considered cryptographically secure.
    elseif ($fh = @fopen('/dev/urandom', 'rb')) {
      // PHP only performs buffered reads, so in reality it will always read
      // at least 4096 bytes. Thus, it costs nothing extra to read and store
      // that much so as to speed any additional invocations.
      $bytes .= fread($fh, max(4096, $missing_bytes));
      fclose($fh);
    }
```

On a computer running Windows without the OpenSSL extension installed, `$bytes` won't be initialized and `strlen()` in the next line will get `NULL`.
Since the OpenSSL extension isn't required to install or run Backdrop ([`system_requirements()`](https://docs.backdropcms.org/api/backdrop/core!modules!system!system.install/function/system_requirements/1) doesn't show an error if it's missing), that is a possibility.

**Comment by indigoxela** [comment]
> On a computer running Windows without the OpenSSL extension installed...

Ah, now I get it. In that case neither function openssl_random_pseudo_bytes() exists, nor is '/dev/urandom' available, so `$bytes` is still NULL. Nice catch. :+1: 

It seems OK to cast to string in that rather rare case, but I can't actually test (no Windows).

**Comment by kiamlaluno** [comment]
I don't like much to cast a value just to avoid an exception or a warning, but in this case, since the content of `$bytes` is handled by `backdrop_random_bytes()`, I think it's acceptable.

The alternative would be using code similar to the following one.

```php
if (empty($bytes) || strlen($bytes) < $count) {
  // …
}
```

I used `strlen((string) $bytes)` because that was already used by `backdrop_random_bytes()`, but I can use the alternative I shown here, if wanted.
I personally like the alternative better because it avoids to calculate the length of an empty string and checks its less than `$count`, which is useless, since that is clearly true.

**Comment by indigoxela** [comment]
> I personally like the alternative better because it avoids to calculate ... which is useless

For sure a valid point. If you prefer that - I'm supporting it. :+1:

**Comment by kiamlaluno** [comment]
I changed the code, since in that line there was already a `if` statement.

**Comment by indigoxela** [comment]
This looks great now, we only need some Windows user to test it. :+1:

**Comment by klonos** [comment]
> On a computer running Windows without the OpenSSL extension installed...


> ...we only need some Windows user to test it. 

I've installed Windows 10 x64 on a VirtualBox, set up XAMPP with php 8.1.6, and disabled OpenSSL entirely in apache. I then tested installing Backdrop 1.22.2 and got this:

![image](https://user-images.githubusercontent.com/2423362/185756904-7ced1e64-d432-472c-a5b7-d09b61d412ad.png)

I then applied the changes from the PR manually, refreshed the page, and the error went away. I was able to proceed with the installation without any other errors. This is RTBC 👍🏼 

Thank you @kiamlaluno 🙏🏼
